dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/dataapproval/DataApprovalPermissionsEvaluator.java

```diff
@@ -97,9 +95,18 @@ public static void invalidateCache()
      */
     public static DataApprovalPermissionsEvaluator makePermissionsEvaluator( CurrentUserService currentUserService,
             OrganisationUnitService organisationUnitService, SystemSettingManager systemSettingManager,
-            DataApprovalLevelService dataApprovalLevelService )
+            DataApprovalLevelService dataApprovalLevelService, CacheProvider cacheProvider )
     {
         DataApprovalPermissionsEvaluator ev = new DataApprovalPermissionsEvaluator();
+        
+        if ( USER_APPROVAL_LEVEL_CACHE == null )
```

Comment:
This part looks a bit prone to concurrency issues. I.e. theoretically there is a gap between the null check and the instantiation of the cache where two threads can overwrite each other. (Perhaps this is theoretical but still worth adjusting in my mind). I think either i) expose the cache itself in the constructor, and pass it into the class or ii) make the cache non-static and part of the class and create it normally in the constructor (and live with the fact that the cache is not shared among threads).

This code design is not optimal in the first place of course.